### ISSUE TITLE
Issue with missing buttons in game. About aspect ratio

### ISSUE THREAD

**Issue by M1sh2** [nipkownix/re4_tweaks#54]
The problem is very old (from the release), it seems to be somehow related to the language that is used in the system. I noticed that it manifests itself if you start using, for example, the Russian language, the Taiwanese language, perhaps some other. The buttons disappear completely, although I use the English version, they are not there when the context menu appears.
Screenshots are attached.
And also, is it possible to correct aspect ratio for 4:3, 5:4, and for other non-standard resolutions? The old permissions even at that time were greatly curtailed by cinematic black bars.
![20220202181359_1](https://user-images.githubusercontent.com/59886778/152146097-8f981959-b064-4297-86d5-bd831860f665.jpg) 
![20220202181532_1](https://user-images.githubusercontent.com/59886778/152146479-317f3e2f-972e-495f-8810-17470e14e0ad.jpg)
![20220202181801_1](https://user-images.githubusercontent.com/59886778/152147604-67b42d6f-845d-4ba5-9e25-188600945667.jpg)
![20220202181601_1](https://user-images.githubusercontent.com/59886778/152147682-9361f5ba-93c4-4dbe-a034-6fed5ce203bc.jpg)

**Comment by emoose** [comment]
Hmm that is weird - I remember something about C functions like printf relying on locale, causing stuff like %f to show as eg. 3.1415 in one locale while 3,1415 in others - wonder if overriding C locale on startup might help at all.

**Comment by RiasatSalminSami** [comment]
If I run the game with MSAA, I can see that the game is rendered behind the black bars during both gameplay and cutscenes while playing on 4:3 aspect ratio.

That means if we can somehow remove the black bars from top and bottom on 4:3, it should work. Although it might show non-animated stuff during cutscenes behind those black bar areas, but gameplay should be fine.

**Comment by nipkownix** [comment]
>E: tried using a hook to call games setlocale func before running WinMain, setlocale(LC_ALL, "ru-RU") didn't seem to have any effect unfortunately.

Might be because you actually need to have that language installed on windows, no? (not sure about how locale works on Windows, but that is how it works on Linux at least).

>That means if we can somehow remove the black bars from top and bottom on 4:3, it should work. Although it might show non-animated stuff during cutscenes behind those black bar areas, but gameplay should be fine.

Yeah, removing the black bars causes weird effects like that.
Also, IIRC, some of the game post processing effects is hardcoded to not go over the black bars, so you'd end up with some of those effects not reaching the top/bottom of the screen.

I've also tried just cutting the sides of the screen and adjusting the HUD's position. It does work fine, but the menus have been designed for 16:9, so we end up cutting some necessary stuff from them.

We _could_ make it so that only gameplay has the zooming + cutting the sides, while the interface itself reverts to the usual black bars behavior. Haven't tested that yet.

**Comment by RiasatSalminSami** [comment]
Yeah, shaving off horizontal view and zooming in during only gameplay segments and leaving the menus intact in 16:9 seems to be the best solution in this case.

Although it would be interesting to see how those post process effects look with black bars removed. If you try it out, do share some screenshots, I'd love to see those.

**Comment by emoose** [comment]
> Might be because you actually need to have that language installed on windows, no? (not sure about how locale works on Windows, but that is how it works on Linux at least).

Oh good point, well installing Russian locale + switching to it does make buttons disappear for me, but if I switch back to English & try setlocale again it still has no effect, too bad.

**Comment by nipkownix** [comment]
>If anyone knows the file that contains button images inside that might help to track down what code is using it.

07000000.pack.lfs
Texture number 0332 inside that pack.

Many things are stored inside 07000000.pack.lfs though, so I'm not sure how helpful that is going to be :/

**Comment by emoose** [comment]
Oh I'm using Russian keyboard too, didn't think of that... looking at game it seems to use `GetKeyboardLayout` in a few different places, maybe keys actually chance on certain layouts, but they didn't handle some properly? Does anyone know if there's a layout that does change the key images shown?

**Comment by nipkownix** [comment]
>Also @nipkownix does re4_tweaks have anything for IAT hooking? Could be easiest to hook the GetKeyboardLayout IAT itself instead of all the calls to it (10+)

Hmm, not really, but I've used the code from this PR recently and is seems to work pretty nicely:

https://github.com/ThirteenAG/WidescreenFixesPack/pull/1045/files#diff-45b7d371439728c02c91a7336f47a4c2e580d78cc0e840ba12838c485849237e

**Comment by emoose** [comment]
Seems it is `GetKeyboardLayout` related, at least I tried hooking the `ToAsciiEx` calls that are used after it and made them call `ToUnicodeEx` instead, which gave some kind of result:

![20220203000422_1](https://user-images.githubusercontent.com/1755499/152258463-b7a60e45-5532-41e3-a85b-259a4143833b.jpg)

Unfortunately 0 doesn't seem to be the correct key for russian KB, should be У or у - guess they might not have a font to display russian text?

We could probably just override `GetKeyboardLayout` to return a normal english layout, but if there are layouts that can actually display fine we might need to check for those, or make it switch to english layout on certain unsupported ones.

**Comment by nipkownix** [comment]
These are all the available textures for prompts:

https://i.imgur.com/M55KZ7i.jpg

Doesn't look like there's anything available for Russian.

**Comment by M1sh2** [comment]
Yes, the decision to set English buttons by default will be correct in this situation. English and Russian letters are on the keyboards.

**Comment by emoose** [comment]
@M1sh2 could you check whether https://github.com/nipkownix/re4_tweaks/pull/64 fixes it for you when using Russian layout?
